이미지url에 id 대신 파일 이름을 사용하도록 변경

kwonyj1022 opened the pull request:
<!-- 반드시 Backend/Androiod 라벨과 리뷰어를 등록해주세요! -->

## 📄 작업 내용 요약
- 이미지 url에 id 대신 파일 이름을 사용하도록 변경
- 이미지 이름 사용으로 인한 n+1 개선
- 이미지 파일 조회 시 이미지 이름 기반의 url로 통신하도록 api 수정

## 🙋🏻 리뷰 시 주의 깊게 확인해야 하는 코드
- file changes가 많지만, ImageCalculator.calculateBy()의 파라미터와 반환값 타입이 달라져서 그렇습니다.
- n+1 개선 부분은 도메인별로 커밋을 나눠두었습니다.
- cloudfront 활용해서 s3에서 조회하는 것은 캠퍼스 가서 aws 접속해서 좀 더 알아보고 추가 또는 별도의 pr로 만들겠습니다.

## 📎 Issue 번호
- closed #715

kwonyj1022 committed:
refactor: 이미지 상대 url 기반으로 url을 계산하는 메서드 파라미터로 이미지 id 대신 이미지 이름을 받도록 변경

kwonyj1022 committed:
refactor: 이미지 절대 url 기반으로 url을 계산하는 메서드 파라미터로 이미지 id 대신 이미지 이름을 받도록 변경

kwonyj1022 committed:
refactor: 경매 조회 시 판매자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: 입찰 목록 조회 시 입찰자 정보 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: 채팅방 관련 조회 시 참여자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: qna 관련 조회 시 사용자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: 신고 관련 조회 시 사용자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: 평가 관련 조회 시 사용자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: 사용자 정보 조회 시 사용자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
refactor: 메시지 알림 시 작성자 프로필 이미지 조회로 인한 n+1 개선

kwonyj1022 committed:
feat: 이미지 이름이 포함된 url로 조회 시 이미지 파일을 반환하는 기능 추가

kwonyj1022 committed:
docs: 문서 최신화

JJ503 commented on backend/ddang/src/main/java/com/ddang/ddang/chat/infrastructure/persistence/QuerydslChatRoomAndImageRepository.java:
```diff
@@ -25,17 +25,18 @@ public Optional<ChatRoomAndImageDto> findChatRoomById(final Long chatRoomId) {
-                            .leftJoin(chatRoom.buyer).fetchJoin()
-                            .leftJoin(chatRoom.auction, auction).fetchJoin()
-                            .leftJoin(auction.seller).fetchJoin()
+                            .join(chatRoom.buyer).fetchJoin()
+                            .join(chatRoom.auction, auction).fetchJoin()
+                            .join(auction.seller).fetchJoin()
+                            .leftJoin(auction.seller.profileImage).fetchJoin()
                             .leftJoin(auctionImage).on(auctionImage.id.eq(
```
### 칭찬
👍👍👍

JJ503 commented on backend/ddang/src/main/java/com/ddang/ddang/image/domain/ProfileImage.java:
```diff
@@ -19,8 +19,6 @@
 public class ProfileImage {
 
     public static final String DEFAULT_PROFILE_IMAGE_STORE_NAME = "default_profile_image.png";
-    // TODO: 10/13/23 앞으로 id가 아닌 store name으로 진행하기로 했는데, 임시로 해둡니다. 추후 삭제해주시면 감사하겠습니다.
-    public static final String DEFAULT_PROFILE_IMAGE_ID = "1";
```
### 칭찬
잊지않고 해주셨군요 👍

apptie replied:
### 칭찬

훌륭합니다 👍

JJ503 commented on backend/ddang/src/main/java/com/ddang/ddang/qna/infrastructure/JpaQuestionRepository.java:
```diff
@@ -14,10 +14,12 @@ public interface JpaQuestionRepository extends JpaRepository<Question, Long> {
         FROM Question q
-        JOIN FETCH q.writer
+        JOIN FETCH q.writer w
+        LEFT JOIN FETCH w.profileImage
         LEFT JOIN FETCH q.answer
         JOIN FETCH q.auction a
-        JOIN FETCH a.seller
+        JOIN FETCH a.seller s
+        JOIN FETCH s.profileImage
```
### 칭찬
storename으로 변경되면서 fetch join을 해줘야 하게 됐군요..!
수정하시느라 고생 많으셨습니다 👍👍👍

swonny replied:
### 질문

제이미가 남겨주신 칭찬에 대한 질문입니다!!
이해한 게 맞는지 궁금해서.. 남겨요
storename은 id가 아니고, id가 아닌 다른 필드를 조회하기 위해 추가쿼리가 발생하기 때문에 fetch join을 해야한다고 이해했는데 맞나요?!

근데 엔초 진짜 꼼꼼하고 똑똑하네요

kwonyj1022 replied:
넵 맞습니다!!

JJ503 commented on backend/ddang/src/main/java/com/ddang/ddang/image/infrastructure/persistence/JpaAuctionImageRepository.java:
```diff
@@ -2,6 +2,14 @@
 
 public interface JpaAuctionImageRepository extends JpaRepository<AuctionImage, Long> {
+
+    @Query("""
+        SELECT COUNT(auction_image) > 0
+        FROM AuctionImage auction_image
+        WHERE auction_image.image.storeName = :storeName
+    """)
+    boolean existsByStoreName(final String storeName);
```
### 질문
네이밍만으로 충분해보이는데 `@Query`로 해줘야 하나요?

kwonyj1022 replied:
넵 AuctionImage 안에 Image가 임베디드로 있어서 JPQL 없이 메서드 네이밍만으로는 불가능합니다.

JJ503 reviewed (approved):
엔초 자잘한 수정 사항이 많았을텐데 정말 고생 많으셨습니다!
제 눈에 별다른 수정 사항은 보이지 않아 approve 합니다.
다만, 궁금한 점이 있어 질문 남겨두었으니 해당 부분만 확이해주시면 감사하겠습니다.

apptie reviewed (approved):
문제가 없어 보이므로 approve 하겠습니다 
제이미의 문서 최신화랑 충돌날 것 같은데 문서 충돌이야 뭐...괜찮겠죠

swonny commented on backend/ddang/src/main/java/com/ddang/ddang/user/domain/repository/UserRepository.java:
```diff
@@ -9,6 +9,7 @@ public interface UserRepository {
     User save(final User user);
 
     Optional<User> findById(final Long id);
+    Optional<User> findByIdWithProfileImage(final Long id);
```
### 필수

개행이 한 줄 필요해보입니당

swonny commented on backend/ddang/src/main/java/com/ddang/ddang/image/presentation/ImageController.java:
```diff
@@ -12,25 +11,27 @@
-    public ResponseEntity<Resource> downloadProfileImage(@PathVariable Long id) throws MalformedURLException {
-        final Resource resource = imageService.readProfileImage(id);
+    @GetMapping("/users/images/{storeName}")
+    public ResponseEntity<Resource> downloadProfileImage(@PathVariable String storeName) throws MalformedURLException {
```
### 필수

final이 누락되어버렸어요..!

swonny commented on backend/ddang/src/main/java/com/ddang/ddang/image/presentation/ImageController.java:
```diff
@@ -12,25 +11,27 @@
-    public ResponseEntity<Resource> downloadAuctionImage(@PathVariable Long id) throws MalformedURLException {
-        final Resource resource = imageService.readAuctionImage(id);
+    @GetMapping("/auctions/images/{storeName}")
+    public ResponseEntity<Resource> downloadAuctionImage(@PathVariable String storeName) throws MalformedURLException {
```
### 필수

final이 누락되어버렸어요..! 22

swonny commented on backend/ddang/src/test/java/com/ddang/ddang/image/infrastructure/persistence/JpaAuctionImageRepositoryTest.java:
```diff
@@ -37,27 +30,25 @@ class JpaAuctionImageRepositoryTest extends JpaAuctionImageRepositoryFixture {
         assertThat(actual.getId()).isPositive();
     }
 
+
```
### 필수

이번엔 개행이 두 번 추가되었어요!

swonny reviewed (approved):
엔초 진짜 자잘한 수정사항이 많았는데 역시 꼼꼼하게 잘 해주셨군요
감사합니다
고생하셨습니다!!

kwonyj1022 committed:
style: 개행 수정 및 누락된 final 추가

kwonyj1022 committed:
ci: 브랜치 최신화

# Conflicts:
#	backend/ddang/src/main/java/com/ddang/ddang/user/presentation/dto/ReadUserResponse.java
#	backend/ddang/src/main/java/com/ddang/ddang/user/presentation/dto/response/ReadAuctionDetailResponse.java
#	backend/ddang/src/main/java/com/ddang/ddang/user/presentation/dto/response/ReadUserResponse.java
#	backend/ddang/src/main/resources/static/docs/docs.html

kwonyj1022 committed:
rename: test에 사용되는 fixture 클래스 이름 변경

swonny reviewed (approved):
확인 완료했어요~~ 고생하셨습니다 엔초

kwonyj1022 committed:
ci: 브랜치 최신화

# Conflicts:
#	backend/ddang/src/main/java/com/ddang/ddang/chat/infrastructure/persistence/QuerydslChatRoomAndMessageAndImageRepository.java
#	backend/ddang/src/main/java/com/ddang/ddang/user/domain/repository/UserRepository.java
#	backend/ddang/src/main/java/com/ddang/ddang/user/infrastructure/persistence/UserRepositoryImpl.java
#	backend/ddang/src/test/java/com/ddang/ddang/chat/presentation/fixture/ChatRoomControllerFixture.java

kwonyj1022 committed:
ci: 충돌 해결
